session/pessimistic_test.go

```diff
@@ -927,6 +928,46 @@ func (s *testPessimisticSuite) TestPessimisticReadCommitted(c *C) {
 	tk.MustExec("commit;")
 }
 
+func (s *testPessimisticSuite) TestPessimisticLockNonExistsKey(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk1 := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (k int primary key, c int)")
+	tk.MustExec("insert t values (1, 1), (3, 3), (5, 5)")
+
+	// verify that select with project and filter on a non exists key still locks the key.
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert t values (8, 8)") // Make the transaction dirty.
+	tk.MustQuery("select c + 1 from t where k = 2 and c = 2 for update").Check(testkit.Rows())
```

Comment:
if we change this sql to 

`select c + 1 from t where k = (2, 3) and c = 2 for update`

`(3, 3)` also be locked after this pr, but mysql seems do the same things - -

Reply:
Yes, it is expected, we lock it even if it doesn't pass the filter.